### PR TITLE
[AIR/Docs] Large model inference examples

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR adds two simple examples showing how to do batch inference with Ray Data using Stable Diffusion and GPT-J models.

Note that the datasets used are very small, allowing them to be ran quickly by both the user and CI. Examples with large scale datasets (>=1 TB) are left as a TODO.

DeepSpeed Inference example will be added in a separate PR.

We are using Ray Data instead of `BatchPredictor` due to https://github.com/ray-project/ray/issues/32923.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/examples/BUILD [hunk 1]
```diff
@@ -46,6 +46,8 @@ py_test_run_all_notebooks(
         "feast_example.ipynb",  # REGRESSION
         "upload_to_comet_ml.ipynb",  # Needs credentials
         "upload_to_wandb.ipynb",  # Needs credentials
+        "gptj_batch_prediction.ipynb",  # Requires GPUs
+        "stablediffusion_batch_prediction.ipynb",  # Requires GPUs
```
Comment (richardliaw) [inline doc/source/ray-air/examples/BUILD:50]: shouldn't this go into the GPU build?

Comment (Yard1) [reply doc/source/ray-air/examples/BUILD:50]: this is the exclude list for the CPU build :)

**Review by ericl — COMMENTED** [review]
Looks good at a high level. Should we add a TODO to add a larger-scale example? This one is quite small.

**Comment by Yard1** [comment]
@ericl I think it's fine to keep them small - after all, it is *that* easy. The DeepSpeed example I am working on will be bigger and more advanced, and the same goes for fine-tuning examples that are next in line.

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "preds = ds.map_batches(\n",
+    "    PredictCallable,\n",
+    "    batch_size=4,\n",
+    "    fn_constructor_kwargs=dict(model_id=model_id, revision=revision),\n",
+    "    compute=\"actors\",\n",
+    "    num_gpus=1,\n",
```
Comment (ericl) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:155]: Shall we also mention this also works with num_gpus=2 or more?

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    "    def __init__(self, model_id: str, revision: str = None):\n",
+    "        from transformers import AutoModelForCausalLM, AutoTokenizer\n",
+    "        import torch\n",
+    "\n",
+    "        self.model = AutoModelForCausalLM.from_pretrained(\n",
+    "            model_id,\n",
+    "            revision=revision,\n",
+    "            torch_dtype=torch.float16,\n",
+    "            low_cpu_mem_usage=True,\n",
+    "            device_map=\"auto\",\n",
```
Comment (ericl) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:112]: comment that auto enables model parallel support on gpus on one node

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 4]
```diff
@@ -0,0 +1,224 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# GPT-J6-B Batch Prediction with Ray AIR\n",
+    "\n",
+    "In this example, we will showcase how to use the Ray AIR for **GPT-J batch inference**. GPT-J is a GPT-2-like causal language model trained on the Pile dataset. This particular model has 6 billion parameters.\n",
```
Comment (richardliaw) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:10]: Can you add some tip about, they should first read about Key Concepts from AIR and Data?

Comment (richardliaw) [reply doc/source/ray-air/examples/gptj_batch_prediction.ipynb:10]: like "Prerequisite Material"

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+   "source": [
+    "import ray"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We define a runtime environment to ensure that the Ray workers have access to all the necessary packages."
```
Comment (richardliaw) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:48]: Also add, you don't need to do this if you already have packages installed.

Also link to runtime env docs

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    "\n",
+    "ds = ray.data.from_pandas(pd.DataFrame([prompt] * 10, columns=[\"prompt\"]))"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You may notice that we are not using an AIR {class}`Predictor <ray.train.predictor.Predictor>` here. Predictors are mainly intended to be used with AIR {class}`Checkpoints <ray.air.Checkpoint>`. Since we will be using a pretrained model from Hugging Face hub, it is simpler to instead use {meth}`map_batches <ray.data.Dataset.map_batches>` with a [callable class UDF](transform_datasets_callable_classes). This will allow us to save time by initializing a model just once and then feed it multiple batches of data."
```
Comment (richardliaw) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:93]: As a new user, I think this "Predictor" part is a bit confusing. Maybe leave it for the end.

**Review by richardliaw — COMMENTED** [review]
all same comments apply across notebooks. nice work!

File: doc/source/ray-air/examples/gptj_batch_prediction.ipynb [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since we will be using a pretrained model from Hugging Face hub, the simplest way is to use {meth}`map_batches <ray.data.Dataset.map_batches>` with a [callable class UDF](transform_datasets_callable_classes). This will allow us to save time by initializing a model just once and then feed it multiple batches of data.\n",
+    "\n",
+    "You may notice that we are not using an AIR {class}`Predictor <ray.train.predictor.Predictor>` here. This is because Predictors are mainly intended to be used with AIR {class}`Checkpoints <ray.air.Checkpoint>`, which we don't for this example."
```
Comment (richardliaw) [inline doc/source/ray-air/examples/gptj_batch_prediction.ipynb:97]: sorry, can you move this to the bottom of the doc?

**Review by richardliaw — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Pull request merged**